### ISSUE TITLE
Xmodmap/"caps lock to hyper" rebinding is undetected

### ISSUE THREAD

**Issue by rien333** [eXeC64/imv#170]
Although I've actually bound caps lock to the hyper key in `.Xmodmap`,   for some reason `imv` still detects my caps lock as a physical caps lock key. Pretty annoying, because most regular keyboard shortcuts fail to work when capital letters are inputted. I use the physical caps lock as my main modifier to manage my wm (as others have recommend me). I've mapped my caps lock to the hyper key in my `.Xmodmap` like so:

```
# 66 is the caps lock key code, iirc
keycode 66 = Hyper_L
add mod3 = Hyper_L
```

This definitely seems like a bug in `imv` or one of it's dependencies, because:

1. No app, whatever DE (Gnome, KDE, QT, bare-bones x11 apps), actually triggers my  caps lock. I have no idea how `imv` even manages to detect a caps lock press, I suppose it has to bypass x11 in some way to even achieve this. 
2. When caps lock is  actually turned on, a little LED on the key should light up. This doesn't happen when I press caps lock in `imv`.
3. When `imv` thinks caps lock is on, none of my other apps think it's turned on. It's completely confined to `imv`.  

Could it be that `imv` is somehow unaware of the modifiers defined in `.Xmodmap`?

**Comment by eXeC64** [comment]
At present, keyboard layout detection is only working in Wayland, not X11. It's a todo to fix it before v4. Which are you using?

**Comment by rien333** [comment]
X11, so that might make sense I guess.

**Comment by eXeC64** [comment]
Right. I've yet to find a decent way to get the keymap description out of X11 in a nice form that I can pass to xkb. The closest I've found is an xkb function that takes an open xcb handle, but I'm currently using libx11 rather than xcb. I'll probably have to rewrite the X11 window implementation using libxcb instead.

In the mean time, you can set the following environment variables that ought to be honoured: They follows the same syntax as running `setxkbmap`: `XKB_DEFAULT_RULES`, `XKB_DEFAULT_MODEL`, `XKB_DEFAULT_LAYOUT`, `XKB_DEFAULT_VARIANT`, `XKB_DEFAULT_OPTIONS`

**Comment by eXeC64** [comment]
I'm going to close this as a duplicate of #153

**Comment by rien333** [comment]
Whatever xmodmap is doing might bypass any `setxkbmap` configuration. From the [arch wiki](https://wiki.archlinux.org/index.php/Xmodmap):
> xmodmap is _not_ directly related to X keyboard extension (XKB), as it uses different (pre-XKB) ideas on how keycodes are processed within X. 

Just a heads up, I happened to stumble upon this because I don't know a lot about xkb, and never used `setxkbmap`. The keyboard-related rewrite you propose will likely fix both my problem and #153, but they could be unrelated. Because xmodmap seems relatively independent of the keyboard layout (it's just a key mapping mechanism), I'm not sure if the environment variables you mention are gonna help. I'll do some more research though, or just await your x11 keyboard input rewrite.

**Comment by eXeC64** [comment]
I'd generally recommend migrating over from xmodmap to xkb, as it should be equally powerful while being compatible with a future move to Wayland.

My configuration looks like:

    xkb_layout gb                                                                 
    xkb_options caps:escape,compose:ralt

Which selects the UK keyboard layout, binds my capslock to escape, and makes my right alt key act as a compose key.

I suspect this is probably what you'd want:

    xkb_options caps:Hyper
